File: awiki/python-client.backup.20260308/scripts/local_store.py
"""SQLite local storage for messages and contacts.

[INPUT]: SDKConfig (data_dir for database path)
[OUTPUT]: get_connection(), ensure_schema(), store_message(), store_messages_batch(),
         make_thread_id(), upsert_contact(), execute_sql()
[POS]: Persistence layer — single shared SQLite database for offline message storage and contact management

[PROTOCOL]:
1. Update this header when logic changes
2. Check the folder's CLAUDE.md after updating
"""

from __future__ import annotations

import re
import sqlite3
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from utils.config import SDKConfig

# Current schema version (bump when schema changes)
_SCHEMA_VERSION = 2

# SQL patterns that are always forbidden
_FORBIDDEN_PATTERNS = [
    re.compile(r"\bDROP\b", re.IGNORECASE),
    re.compile(r"\bTRUNCATE\b", re.IGNORECASE),
]

# DELETE without WHERE is forbidden
_DELETE_NO_WHERE = re.compile(
    r"\bDELETE\s+FROM\s+\S+\s*$", re.IGNORECASE,
)


def get_connection() -> sqlite3.Connection:
    """Open (or create) the shared SQLite database.

    Path: <DATA_DIR>/database/awiki.db
    Uses WAL mode for concurrent read/write.

    Returns:
        sqlite3.Connection with check_same_thread=False.
    """
    config = SDKConfig()
    db_dir = config.data_dir / "database"
    db_dir.mkdir(parents=True, exist_ok=True)

    db_path = db_dir / "awiki.db"
    conn = sqlite3.connect(str(db_path), check_same_thread=False)
    conn.execute("PRAGMA journal_mode=WAL")
    conn.execute("PRAGMA foreign_keys=ON")
    conn.row_factory = sqlite3.Row
    return conn


def ensure_schema(conn: sqlite3.Connection) -> None:
    """Create tables, views, and indexes if they don't exist.

    Idempotent — safe to call on every connection open.
    Handles migration from schema v1 (adds credential_name column).
    """
    version = conn.execute("PRAGMA user_version").fetchone()[0]
    if version >= _SCHEMA_VERSION:
        return

    conn.executescript("""
        CREATE TABLE IF NOT EXISTS contacts (
            did             TEXT PRIMARY KEY,
            name            TEXT,
            handle          TEXT,
            nick_name       TEXT,
            bio             TEXT,
            profile_md      TEXT,
            tags            TEXT,
            relationship    TEXT,
            first_seen_at   TEXT,
            last_seen_at    TEXT,
            metadata        TEXT
        );

        CREATE TABLE IF NOT EXISTS messages (
            msg_id          TEXT PRIMARY KEY,
            thread_id       TEXT NOT NULL,
            direction       INTEGER NOT NULL DEFAULT 0,
            sender_did      TEXT,
            receiver_did    TEXT,
            group_id        TEXT,
            group_did       TEXT,
            content_type    TEXT DEFAULT 'text',
            content         TEXT,
            server_seq      INTEGER,
            sent_at         TEXT,
            stored_at       TEXT NOT NULL,
            is_e2ee         INTEGER DEFAULT 0,
            is_read         INTEGER DEFAULT 0,
            sender_name     TEXT,
            metadata        TEXT,
            credential_name TEXT
        );

        CREATE INDEX IF NOT EXISTS idx_messages_thread
            ON messages(thread_id, sent_at);
        CREATE INDEX IF NOT EXISTS idx_messages_direction
            ON messages(direction);
        CREATE INDEX IF NOT EXISTS idx_messages_sender
            ON messages(sender_did);
        CREATE INDEX IF NOT EXISTS idx_messages_credential
            ON messages(credential_name);
    """)

    # Migration: add credential_name column if upgrading from v1
    if version == 1:
        # Check if credential_name column already exists
        columns = {
            row[1]
            for row in conn.execute("PRAGMA table_info(messages)").fetchall()
        }
        if "credential_name" not in columns:
            conn.execute(
                "ALTER TABLE messages ADD COLUMN credential_name TEXT"
            )
            conn.execute(
                "CREATE INDEX IF NOT EXISTS idx_messages_credential "
                "ON messages(credential_name)"
            )

    # Views (DROP + CREATE to allow schema evolution)
    conn.executescript("""
        DROP VIEW IF EXISTS threads;
        CREATE VIEW threads AS
        SELECT
            thread_id,
            COUNT(*)                                        AS message_count,
            SUM(CASE WHEN is_read = 0 AND direction = 0
                     THEN 1 ELSE 0 END)                    AS unread_count,
            MAX(COALESCE(sent_at, stored_at))               AS last_message_at,
            (SELECT m2.content FROM messages m2
             WHERE m2.thread_id = m.thread_id
             ORDER BY COALESCE(m2.sent_at, m2.stored_at) DESC
             LIMIT 1)                                       AS last_content
        FROM messages m
        GROUP BY thread_id;

        DROP VIEW IF EXISTS inbox;
        CREATE VIEW inbox AS
        SELECT * FROM messages WHERE direction = 0
        ORDER BY COALESCE(sent_at, stored_at) DESC;

        DROP VIEW IF EXISTS outbox;
        CREATE VIEW outbox AS
        SELECT * FROM messages WHERE direction = 1
        ORDER BY COALESCE(sent_at, stored_at) DESC;
    """)

    conn.execute(f"PRAGMA user_version = {_SCHEMA_VERSION}")
    conn.commit()


def make_thread_id(
    my_did: str,
    *,
    peer_did: str | None = None,
    group_id: str | None = None,
) -> str:
    """Generate a deterministic thread ID.

    - Private chat: dm:{min_did}:{max_did} (sorted for symmetry)
    - Group chat: group:{group_id}

    Args:
        my_did: Current user's DID.
        peer_did: Peer's DID (for private chats).
        group_id: Group ID (for group chats).

    Returns:
        Thread identifier string.
    """
    if group_id:
        return f"group:{group_id}"
    if peer_did:
        pair = sorted([my_did, peer_did])
        return f"dm:{pair[0]}:{pair[1]}"
    return f"dm:{my_did}:unknown"


def store_message(
    conn: sqlite3.Connection,
    *,
    msg_id: str,
    thread_id: str,
    direction: int,
    sender_did: str,
    content: str,
    receiver_did: str | None = None,
    group_id: str | None = None,
    group_did: str | None = None,
    content_type: str = "text",
    server_seq: int | None = None,
    sent_at: str | None = None,
    is_e2ee: bool = False,
    is_read: bool = False,
    sender_name: str | None = None,
    metadata: str | None = None,
    credential_name: str | None = None,
) -> None:
    """Store a single message. Duplicate msg_id is silently ignored (INSERT OR IGNORE).

    Args:
        conn: SQLite connection.
        msg_id: Unique message identifier.
        thread_id: Thread identifier (from make_thread_id).
        direction: 0 = incoming, 1 = outgoing.
        sender_did: Sender's DID.
        content: Message content text.
        receiver_did: Receiver's DID (optional).
        group_id: Group ID (optional).
        group_did: Group DID (optional).
        content_type: Content type (default: "text").
        server_seq: Server sequence number (optional).
        sent_at: Server-side send timestamp (optional).
        is_e2ee: Whether this message was E2EE encrypted.
        is_read: Whether this message has been read.
        sender_name: Display name of sender (optional).
        metadata: JSON metadata string (optional).
        credential_name: Credential name that owns this message (optional).
    """
    now = datetime.now(timezone.utc).isoformat()
    conn.execute(
        """INSERT OR IGNORE INTO messages
           (msg_id, thread_id, direction, sender_did, receiver_did,
            group_id, group_did, content_type, content, server_seq,
            sent_at, stored_at, is_e2ee, is_read, sender_name, metadata,
            credential_name)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
        (
            msg_id, thread_id, direction, sender_did, receiver_did,
            group_id, group_did, content_type, content, server_seq,
            sent_at, now, int(is_e2ee), int(is_read), sender_name, metadata,
            credential_name,
        ),
    )
    conn.commit()


def store_messages_batch(
    conn: sqlite3.Connection,
    batch: list[dict],
    credential_name: str | None = None,
) -> None:
    """Store multiple messages in a single transaction.

    Each dict in batch should contain the same keyword arguments as store_message().
    Duplicates are silently ignored.

    Args:
        conn: SQLite connection.
        batch: List of message dicts.
        credential_name: Default credential name applied to all messages in batch
            (individual messages can override via "credential_name" key).
    """
    if not batch:
        return

    now = datetime.now(timezone.utc).isoformat()
    rows = []
    for msg in batch:
        rows.append((
            msg.get("msg_id", ""),
            msg.get("thread_id", ""),
            msg.get("direction", 0),
            msg.get("sender_did"),
            msg.get("receiver_did"),
            msg.get("group_id"),
            msg.get("group_did"),
            msg.get("content_type", "text"),
            msg.get("content", ""),
            msg.get("server_seq"),
            msg.get("sent_at"),
            now,
            int(msg.get("is_e2ee", False)),
            int(msg.get("is_read", False)),
            msg.get("sender_name"),
            msg.get("metadata"),
            msg.get("credential_name", credential_name),
        ))

    conn.executemany(
        """INSERT OR IGNORE INTO messages
           (msg_id, thread_id, direction, sender_did, receiver_did,
            group_id, group_did, content_type, content, server_seq,
            sent_at, stored_at, is_e2ee, is_read, sender_name, metadata,
            credential_name)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
        rows,
    )
    conn.commit()


def upsert_contact(
    conn: sqlite3.Connection,
    *,
    did: str,
    **fields: Any,
) -> None:
    """Insert or update a contact record.

    Args:
        conn: SQLite connection.
        did: Contact's DID (primary key).
        **fields: Optional fields to set (name, handle, nick_name, bio, etc.).
    """
    now = datetime.now(timezone.utc).isoformat()

    # Check if contact exists
    existing = conn.execute(
        "SELECT did FROM contacts WHERE did = ?", (did,),
    ).fetchone()

    allowed_fields = {
        "name", "handle", "nick_name", "bio", "profile_md",
        "tags", "relationship", "first_seen_at", "last_seen_at", "metadata",
    }
    filtered = {k: v for k, v in fields.items() if k in allowed_fields and v is not None}

    if existing:
        if filtered:
            filtered["last_seen_at"] = now
            set_clause = ", ".join(f"{k} = ?" for k in filtered)
            values = list(filtered.values()) + [did]
            conn.execute(
                f"UPDATE contacts SET {set_clause} WHERE did = ?", values,
            )
    else:
        filtered.setdefault("first_seen_at", now)
        filtered.setdefault("last_seen_at", now)
        filtered["did"] = did
        columns = ", ".join(filtered.keys())
        placeholders = ", ".join("?" for _ in filtered)
        conn.execute(
            f"INSERT INTO contacts ({columns}) VALUES ({placeholders})",
            list(filtered.values()),
        )

    conn.commit()


def execute_sql(
    conn: sqlite3.Connection,
    sql: str,
    params: tuple = (),
) -> list[dict[str, Any]]:
    """Execute a SQL statement with safety checks.

    Allowed: SELECT, INSERT, UPDATE, DELETE (with WHERE), REPLACE, ALTER, CREATE.
    Forbidden: DROP, TRUNCATE, DELETE without WHERE, multiple statements.

    Args:
        conn: SQLite connection.
        sql: SQL statement to execute.
        params: Query parameters.

    Returns:
        For SELECT: list of row dicts.
        For write operations: [{"rows_affected": N}].

    Raises:
        ValueError: If the SQL violates safety rules.
    """
    stripped = sql.strip().rstrip(";")

    # Reject multiple statements
    if ";" in stripped:
        raise ValueError("Multiple statements are not allowed")

    # Reject forbidden patterns
    for pattern in _FORBIDDEN_PATTERNS:
        if pattern.search(stripped):
            raise ValueError(f"Forbidden SQL operation: {pattern.pattern}")

    # Reject DELETE without WHERE
    if re.match(r"\s*DELETE\b", stripped, re.IGNORECASE):
        if not re.search(r"\bWHERE\b", stripped, re.IGNORECASE):
            raise ValueError("DELETE without WHERE clause is not allowed")

    cursor = conn.execute(stripped, params)

    if stripped.upper().lstrip().startswith("SELECT"):
        columns = [desc[0] for desc in cursor.description] if cursor.description else []
        return [dict(zip(columns, row)) for row in cursor.fetchall()]

    conn.commit()
    return [{"rows_affected": cursor.rowcount}]

File: awiki/nodejs-client/scripts/utils/local_store.js
/**
 * SQLite local storage for messages, contacts, and E2EE outbox state.
 *
 * [INPUT]: SDKConfig (data_dir for database path), credential_store (owner DID lookup
 *          during migration)
 * [OUTPUT]: get_connection(), ensure_schema(), store_message(), store_messages_batch(),
 *           queue_e2ee_outbox(), mark_e2ee_outbox_sent(), mark_e2ee_outbox_failed(),
 *           list_e2ee_outbox(), get_e2ee_outbox(), get_message_by_id(), make_thread_id(),
 *           upsert_contact(), rebind_owner_did(), clear_owner_e2ee_data(), execute_sql()
 * [POS]: Persistence layer — single shared SQLite database for offline message storage,
 *        contact management, and resendable E2EE outbox tracking with explicit
 *        owner_did isolation for multi-identity local environments
 *
 * [PROTOCOL]:
 * 1. Update this header when logic changes
 * 2. Check the folder's CLAUDE.md after updating
 */

import Database from 'better-sqlite3';
import path from 'path';
import fs from 'fs';
import { createSDKConfig } from './config.js';

const _SCHEMA_VERSION = 6;

const _V6_TABLES_SQL = `
    CREATE TABLE IF NOT EXISTS contacts (
        owner_did       TEXT NOT NULL DEFAULT '',
        did             TEXT NOT NULL,
        name            TEXT,
        handle          TEXT,
        nick_name       TEXT,
        bio             TEXT,
        profile_md      TEXT,
        tags            TEXT,
        relationship    TEXT,
        first_seen_at   TEXT,
        last_seen_at    TEXT,
        metadata        TEXT,
        PRIMARY KEY (owner_did, did)
    );

    CREATE TABLE IF NOT EXISTS messages (
        msg_id          TEXT NOT NULL,
        owner_did       TEXT NOT NULL DEFAULT '',
        thread_id       TEXT NOT NULL,
        direction       INTEGER NOT NULL DEFAULT 0,
        sender_did      TEXT,
        receiver_did    TEXT,
        group_id        TEXT,
        group_did       TEXT,
        content_type    TEXT DEFAULT 'text',
        content         TEXT,
        title           TEXT,
        server_seq      INTEGER,
        sent_at         TEXT,
        stored_at       TEXT NOT NULL,
        is_e2ee         INTEGER DEFAULT 0,
        is_read         INTEGER DEFAULT 0,
        sender_name     TEXT,
        metadata        TEXT,
        credential_name TEXT NOT NULL DEFAULT '',
        PRIMARY KEY (msg_id, owner_did)
    );

    CREATE TABLE IF NOT EXISTS e2ee_outbox (
        outbox_id            TEXT PRIMARY KEY,
        owner_did            TEXT NOT NULL DEFAULT '',
        peer_did             TEXT NOT NULL,
        session_id           TEXT,
        original_type        TEXT NOT NULL DEFAULT 'text',
        plaintext            TEXT NOT NULL,
        local_status         TEXT NOT NULL DEFAULT 'queued',
        attempt_count        INTEGER NOT NULL DEFAULT 0,
        sent_msg_id          TEXT,
        sent_server_seq      INTEGER,
        last_error_code      TEXT,
        retry_hint           TEXT,
        failed_msg_id        TEXT,
        failed_server_seq    INTEGER,
        metadata             TEXT,
        last_attempt_at      TEXT,
        created_at           TEXT NOT NULL,
        updated_at           TEXT NOT NULL,
        credential_name      TEXT NOT NULL DEFAULT ''
    );
`;

const _V6_INDEX_STATEMENTS = {
    idx_contacts_owner: `
        CREATE INDEX IF NOT EXISTS idx_contacts_owner
            ON contacts(owner_did, last_seen_at DESC)
    `,
    idx_messages_owner_thread: `
        CREATE INDEX IF NOT EXISTS idx_messages_owner_thread
            ON messages(owner_did, thread_id, sent_at)
    `,
    idx_messages_owner_direction: `
        CREATE INDEX IF NOT EXISTS idx_messages_owner_direction
            ON messages(owner_did, direction)
    `,
    idx_messages_owner_sender: `
        CREATE INDEX IF NOT EXISTS idx_messages_owner_sender
            ON messages(owner_did, sender_did)
    `,
    idx_messages_owner: `
        CREATE INDEX IF NOT EXISTS idx_messages_owner
            ON messages(owner_did)
    `,
    idx_messages_credential: `
        CREATE INDEX IF NOT EXISTS idx_messages_credential
            ON messages(credential_name)
    `,
    idx_e2ee_outbox_owner_status: `
        CREATE INDEX IF NOT EXISTS idx_e2ee_outbox_owner_status
            ON e2ee_outbox(owner_did, local_status, updated_at DESC)
    `,
    idx_e2ee_outbox_owner_sent_msg: `
        CREATE INDEX IF NOT EXISTS idx_e2ee_outbox_owner_sent_msg
            ON e2ee_outbox(owner_did, sent_msg_id)
    `,
    idx_e2ee_outbox_owner_sent_seq: `
        CREATE INDEX IF NOT EXISTS idx_e2ee_outbox_owner_sent_seq
            ON e2ee_outbox(owner_did, peer_did, sent_server_seq)
    `,
    idx_e2ee_outbox_credential: `
        CREATE INDEX IF NOT EXISTS idx_e2ee_outbox_credential
            ON e2ee_outbox(credential_name)
    `
};

/**
 * Open (or create) the shared SQLite database.
 * @returns {Database} SQLite database connection
 */
export function get_connection() {
    const config = createSDKConfig();
    const dbDir = path.join(config.data_dir, 'database');
    
    // Create directory if it doesn't exist
    if (!fs.existsSync(dbDir)) {
        fs.mkdirSync(dbDir, { recursive: true });
    }

    const dbPath = path.join(dbDir, 'awiki.db');
    const conn = new Database(dbPath);
    
    // Enable WAL mode and foreign keys
    conn.pragma('journal_mode = WAL');
    conn.pragma('foreign_keys = ON');
    
    console.debug(`Opened local SQLite database path=${dbPath}`);
    return conn;
}

/**
 * Return whether a table exists.
 * @param {Database} conn - Database connection
 * @param {string} tableName - Table name
 * @returns {boolean} True if table exists
 */
function _table_exists(conn, tableName) {
    const stmt = conn.prepare(
        "SELECT name FROM sqlite_master WHERE type='table' AND name = ?"
    );
    const row = stmt.get(tableName);
    return row !== undefined;
}

/**
 * Normalize credential name to string.
 * @param {string|null} credentialName - Credential name
 * @returns {string} Normalized credential name
 */
function _normalize_credential_name(credentialName) {
    return credentialName || 'default';
}

/**
 * Normalize owner DID to string.
 * @param {string|null} ownerDid - Owner DID
 * @returns {string} Normalized owner DID
 */
function _normalize_owner_did(ownerDid) {
    return ownerDid || '';
}

/**
 * Ensure v6 indexes exist.
 * @param {Database} conn - Database connection
 * @returns {string[]} List of created indexes
 */
function _ensure_v6_indexes(conn) {
    const created = [];
    for (const [indexName, sql] of Object.entries(_V6_INDEX_STATEMENTS)) {
        try {
            conn.exec(sql);
            created.push(indexName);
        } catch (error) {
            console.warn(`Failed to create index ${indexName}:`, error.message);
        }
    }
    return created;
}

/**
 * Create schema v6.
 * @param {Database} conn - Database connection
 */
function _create_schema_v6(conn) {
    conn.exec(_V6_TABLES_SQL);
    _ensure_v6_indexes(conn);
}

/**
 * Ensure schema exists.
 * @param {Database} conn - Database connection
 */
export function ensure_schema(conn) {
    if (!_table_exists(conn, 'messages')) {
        _create_schema_v6(conn);
        conn.exec(`PRAGMA user_version = ${_SCHEMA_VERSION}`);
        return;
    }

    // Check schema version
    const stmt = conn.prepare('PRAGMA user_version');
    const result = stmt.get();
    const currentVersion = result.user_version || 0;

    if (currentVersion < _SCHEMA_VERSION) {
        // For now, just ensure indexes exist
        // In a full implementation, we would migrate schema versions
        _ensure_v6_indexes(conn);
        conn.exec(`PRAGMA user_version = ${_SCHEMA_VERSION}`);
    }
}

/**
 * Make a thread ID from two DIDs.
 * @param {string} did1 - First DID
 * @param {string} did2 - Second DID
 * @returns {string} Thread ID
 */
export function make_thread_id(did1, did2) {
    const sorted = [did1, did2].sort();
    return sorted.join('|');
}

/**
 * Store a message in the database.
 * @param {Object} message - Message object
 * @param {string} message.msg_id - Message ID
 * @param {string} message.owner_did - Owner DID
 * @param {string} message.thread_id - Thread ID
 * @param {number} message.direction - Direction (0=incoming, 1=outgoing)
 * @param {string} message.sender_did - Sender DID
 * @param {string} message.receiver_did - Receiver DID
 * @param {string} message.content_type - Content type
 * @param {string} message.content - Message content
 * @param {string} message.title - Message title
 * @param {number} message.server_seq - Server sequence number
 * @param {string} message.sent_at - Sent timestamp
 * @param {number} message.is_e2ee - Is E2EE flag
 * @param {number} message.is_read - Is read flag
 * @param {string} message.sender_name - Sender name
 * @param {Object} message.metadata - Metadata
 * @param {string} credentialName - Credential name
 */
export function store_message(message, credentialName = 'default') {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        const now = new Date().toISOString();
        const stmt = conn.prepare(`
            INSERT OR REPLACE INTO messages (
                msg_id, owner_did, thread_id, direction, sender_did, receiver_did,
                content_type, content, title, server_seq, sent_at, stored_at,
                is_e2ee, is_read, sender_name, metadata, credential_name
            ) VALUES (
                @msg_id, @owner_did, @thread_id, @direction, @sender_did, @receiver_did,
                @content_type, @content, @title, @server_seq, @sent_at, @stored_at,
                @is_e2ee, @is_read, @sender_name, @metadata, @credential_name
            )
        `);
        
        stmt.run({
            ...message,
            stored_at: message.stored_at || now,
            credential_name: _normalize_credential_name(credentialName),
            metadata: message.metadata ? JSON.stringify(message.metadata) : null
        });
    } finally {
        conn.close();
    }
}

/**
 * Store multiple messages in a batch.
 * @param {Object[]} messages - Array of message objects
 * @param {string} credentialName - Credential name
 */
export function store_messages_batch(messages, credentialName = 'default') {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        const now = new Date().toISOString();
        const stmt = conn.prepare(`
            INSERT OR REPLACE INTO messages (
                msg_id, owner_did, thread_id, direction, sender_did, receiver_did,
                content_type, content, title, server_seq, sent_at, stored_at,
                is_e2ee, is_read, sender_name, metadata, credential_name
            ) VALUES (
                @msg_id, @owner_did, @thread_id, @direction, @sender_did, @receiver_did,
                @content_type, @content, @title, @server_seq, @sent_at, @stored_at,
                @is_e2ee, @is_read, @sender_name, @metadata, @credential_name
            )
        `);
        
        const normalizedCredentialName = _normalize_credential_name(credentialName);
        
        conn.transaction(() => {
            for (const message of messages) {
                stmt.run({
                    ...message,
                    stored_at: message.stored_at || now,
                    credential_name: normalizedCredentialName,
                    metadata: message.metadata ? JSON.stringify(message.metadata) : null
                });
            }
        });
    } finally {
        conn.close();
    }
}

/**
 * Queue an E2EE message in the outbox.
 * @param {Object} params - Parameters
 * @param {string} params.outbox_id - Outbox ID
 * @param {string} params.owner_did - Owner DID
 * @param {string} params.peer_did - Peer DID
 * @param {string} params.session_id - Session ID
 * @param {string} params.original_type - Original message type
 * @param {string} params.plaintext - Plaintext message
 * @param {string} credentialName - Credential name
 * @returns {string} Outbox ID
 */
export function queue_e2ee_outbox({
    outbox_id,
    owner_did,
    peer_did,
    session_id,
    original_type = 'text',
    plaintext
}, credentialName = 'default') {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        const now = new Date().toISOString();
        const stmt = conn.prepare(`
            INSERT INTO e2ee_outbox (
                outbox_id, owner_did, peer_did, session_id, original_type, plaintext,
                local_status, attempt_count, created_at, updated_at, credential_name
            ) VALUES (
                @outbox_id, @owner_did, @peer_did, @session_id, @original_type, @plaintext,
                'queued', 0, @created_at, @updated_at, @credential_name
            )
        `);
        
        stmt.run({
            outbox_id,
            owner_did: _normalize_owner_did(owner_did),
            peer_did,
            session_id,
            original_type,
            plaintext,
            created_at: now,
            updated_at: now,
            credential_name: _normalize_credential_name(credentialName)
        });
        
        return outbox_id;
    } finally {
        conn.close();
    }
}

/**
 * Mark an E2EE outbox message as sent.
 * @param {string} outboxId - Outbox ID
 * @param {number} serverSeq - Server sequence number
 * @param {string} sentMsgId - Sent message ID
 */
export function mark_e2ee_outbox_sent(outboxId, serverSeq, sentMsgId) {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        const now = new Date().toISOString();
        const stmt = conn.prepare(`
            UPDATE e2ee_outbox
            SET local_status = 'sent',
                sent_server_seq = @server_seq,
                sent_msg_id = @sent_msg_id,
                updated_at = @updated_at
            WHERE outbox_id = @outbox_id
        `);
        
        stmt.run({
            outbox_id: outboxId,
            server_seq: serverSeq,
            sent_msg_id: sentMsgId,
            updated_at: now
        });
    } finally {
        conn.close();
    }
}

/**
 * Mark an E2EE outbox message as failed.
 * @param {string} outboxId - Outbox ID
 * @param {string} errorCode - Error code
 * @param {string} retryHint - Retry hint
 */
export function mark_e2ee_outbox_failed(outboxId, errorCode, retryHint = null) {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        const now = new Date().toISOString();
        const stmt = conn.prepare(`
            UPDATE e2ee_outbox
            SET local_status = 'failed',
                last_error_code = @last_error_code,
                retry_hint = @retry_hint,
                attempt_count = attempt_count + 1,
                updated_at = @updated_at
            WHERE outbox_id = @outbox_id
        `);
        
        stmt.run({
            outbox_id: outboxId,
            last_error_code: errorCode,
            retry_hint: retryHint,
            updated_at: now
        });
    } finally {
        conn.close();
    }
}

/**
 * Get an E2EE outbox message by ID.
 * @param {string} outboxId - Outbox ID
 * @returns {Object|null} Outbox message or null
 */
export function get_e2ee_outbox(outboxId) {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        const stmt = conn.prepare(`
            SELECT * FROM e2ee_outbox WHERE outbox_id = ?
        `);
        
        const row = stmt.get(outboxId);
        return row || null;
    } finally {
        conn.close();
    }
}

/**
 * List E2EE outbox messages.
 * @param {Object} params - Parameters
 * @param {string} params.owner_did - Owner DID
 * @param {string} params.local_status - Local status filter
 * @param {number} params.limit - Limit
 * @returns {Object[]} Array of outbox messages
 */
export function list_e2ee_outbox({ owner_did, local_status, limit = 50 }) {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        let sql = 'SELECT * FROM e2ee_outbox WHERE owner_did = ?';
        const params = [_normalize_owner_did(owner_did)];
        
        if (local_status) {
            sql += ' AND local_status = ?';
            params.push(local_status);
        }
        
        sql += ' ORDER BY updated_at DESC LIMIT ?';
        params.push(limit);
        
        const stmt = conn.prepare(sql);
        return stmt.all(...params);
    } finally {
        conn.close();
    }
}

/**
 * Get a message by ID.
 * @param {string} msgId - Message ID
 * @param {string} ownerDid - Owner DID
 * @returns {Object|null} Message or null
 */
export function get_message_by_id(msgId, ownerDid) {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        const stmt = conn.prepare(`
            SELECT * FROM messages WHERE msg_id = ? AND owner_did = ?
        `);
        
        const row = stmt.get(msgId, _normalize_owner_did(ownerDid));
        return row || null;
    } finally {
        conn.close();
    }
}

/**
 * Upsert a contact.
 * @param {Object} contact - Contact object
 * @param {string} credentialName - Credential name
 */
export function upsert_contact(contact, credentialName = 'default') {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        const now = new Date().toISOString();
        const stmt = conn.prepare(`
            INSERT OR REPLACE INTO contacts (
                owner_did, did, name, handle, nick_name, bio, profile_md,
                tags, relationship, first_seen_at, last_seen_at, metadata
            ) VALUES (
                @owner_did, @did, @name, @handle, @nick_name, @bio, @profile_md,
                @tags, @relationship, @first_seen_at, @last_seen_at, @metadata
            )
        `);
        
        const normalizedCredentialName = _normalize_credential_name(credentialName);
        const ownerDid = contact.owner_did || '';
        
        stmt.run({
            ...contact,
            owner_did: ownerDid,
            first_seen_at: contact.first_seen_at || now,
            last_seen_at: contact.last_seen_at || now,
            metadata: contact.metadata ? JSON.stringify(contact.metadata) : null
        });
    } finally {
        conn.close();
    }
}

/**
 * Rebind owner DID (for migration).
 * @param {string} oldOwnerDid - Old owner DID
 * @param {string} newOwnerDid - New owner DID
 */
export function rebind_owner_did(oldOwnerDid, newOwnerDid) {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        const stmt = conn.prepare(`
            UPDATE messages
            SET owner_did = @new_owner_did
            WHERE owner_did = @old_owner_did
        `);
        
        stmt.run({
            new_owner_did: newOwnerDid,
            old_owner_did: oldOwnerDid
        });
        
        const stmt2 = conn.prepare(`
            UPDATE contacts
            SET owner_did = @new_owner_did
            WHERE owner_did = @old_owner_did
        `);
        
        stmt2.run({
            new_owner_did: newOwnerDid,
            old_owner_did: oldOwnerDid
        });
        
        const stmt3 = conn.prepare(`
            UPDATE e2ee_outbox
            SET owner_did = @new_owner_did
            WHERE owner_did = @old_owner_did
        `);
        
        stmt3.run({
            new_owner_did: newOwnerDid,
            old_owner_did: oldOwnerDid
        });
    } finally {
        conn.close();
    }
}

/**
 * Clear E2EE data for an owner.
 * @param {string} ownerDid - Owner DID
 */
export function clear_owner_e2ee_data(ownerDid) {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        const stmt = conn.prepare(`
            DELETE FROM e2ee_outbox WHERE owner_did = ?
        `);
        
        stmt.run(_normalize_owner_did(ownerDid));
    } finally {
        conn.close();
    }
}

/**
 * Execute raw SQL (for migrations).
 * @param {string} sql - SQL statement
 * @param {Array} params - Parameters
 * @returns {Object[]} Query results
 */
export function execute_sql(sql, params = []) {
    const conn = get_connection();
    try {
        ensure_schema(conn);
        
        const stmt = conn.prepare(sql);
        return stmt.all(...params);
    } finally {
        conn.close();
    }
}

export default {
    get_connection,
    ensure_schema,
    store_message,
    store_messages_batch,
    queue_e2ee_outbox,
    mark_e2ee_outbox_sent,
    mark_e2ee_outbox_failed,
    get_e2ee_outbox,
    list_e2ee_outbox,
    get_message_by_id,
    make_thread_id,
    upsert_contact,
    rebind_owner_did,
    clear_owner_e2ee_data,
    execute_sql
};
